zhaquirks/tuya/mcu/__init__.py

```diff
@@ -724,3 +730,122 @@ class TuyaLevelControlManufCluster(TuyaMCUCluster):
         17: "_dp_2_attr_update",
         18: "_dp_2_attr_update",
     }
+
+
+class TuyaTemperatureMeasurement(TemperatureMeasurement, TuyaLocalCluster):
+    """Tuya local TemperatureMeasurement cluster."""
+
+
+class TuyaSoilMoisture(SoilMoisture, TuyaLocalCluster):
+    """Tuya local SoilMoisture cluster with a device RH_MULTIPLIER factor if required."""
+
+
+class TuyaQuirkBuilder(QuirkBuilder):
+    """Tuya QuirkBuilder."""
+
+    def __init__(
+        self, manufacturer: str, model: str, registry: DeviceRegistry = _DEVICE_REGISTRY
+    ) -> None:
+        """Init the TuyaQuirkBuilder."""
+        self.tuya_data_point_handlers: dict[int, str] = {}
+        self.tuya_dp_to_attribute: dict[int, DPToAttributeMapping] = {}
+        super().__init__(manufacturer, model, registry)
+
+    def battery(
+        self,
+        dp_id: int,
+        power_cfg: PowerConfiguration = TuyaPowerConfigurationCluster2AAA,
+        scale=2,
+    ) -> QuirkBuilder:
+        """Add a Tuya Battery Power Configuration."""
+        self.tuya_dp(
+            dp_id,
+            power_cfg.ep_attribute,
+            "battery_percentage_remaining",
+            converter=lambda x: x * scale,
+        )
+        self.adds(power_cfg)
+        return self
+
+    def soil_moisture(
+        self,
+        dp_id: int,
+        soil_cfg: TuyaLocalCluster = TuyaSoilMoisture,
+        scale=100,
+    ) -> QuirkBuilder:
+        """Add a Tuya Soil Moisture Configuration."""
+        self.tuya_dp(
+            dp_id,
+            soil_cfg.ep_attribute,
+            "measured_value",
+            converter=lambda x: x * scale,
+        )
+        self.adds(soil_cfg)
+        return self
+
+    def temperature(
+        self,
+        dp_id: int,
+        temp_cfg: TuyaLocalCluster = TuyaTemperatureMeasurement,
+        scale=100,
+    ) -> QuirkBuilder:
+        """Add a Tuya Soil Moisture Configuration."""
+        self.tuya_dp(
+            dp_id,
+            temp_cfg.ep_attribute,
+            "measured_value",
+            converter=lambda x: x * scale,
+        )
+        self.adds(temp_cfg)
+        return self
+
+    def tuya_dp(
+        self,
+        dp_id: int,
+        ep_attribute: str,
+        attribute_name: Union[str, tuple],
+        converter: Optional[
+            Callable[
+                [
+                    Any,
+                ],
+                Any,
+            ]
+        ] = None,
+        dp_converter: Optional[
+            Callable[
+                [
+                    Any,
+                ],
+                Any,
+            ]
+        ] = None,
```

Comment:
Not sure if this works, can we format `converter` and `dp_converter` typing like this?
```python
        converter: Optional[Callable[[Any], Any]] = None,
        dp_converter: Optional[Callable[[Any], Any]] = None,
```

Reply:
Ruff is reformatting it to look like that during pre-commit

Reply:
Went ahead and set ruff to skip that line, I like it better compact.